### PR TITLE
Memory and xr_containers refactor

### PR DESCRIPTION
**Pull request by jazzvaz**

### CODE REVIEW

File: code/engine/xrCore/LzHuf.cpp [hunk 1]
```diff
@@ -93,16 +93,16 @@ class LZfs {
 
     void Init_Output(const int _rsize) {
         // output
-        out_start = (u8*)xr_malloc(_rsize);
+        out_start = new u8[_rsize / sizeof(u8)];
```
Comment (Im-dex) [inline code/engine/xrCore/LzHuf.cpp]: redundant division

File: code/engine/xrCore/LzHuf.cpp [hunk 2]
```diff
@@ -619,7 +619,7 @@ void _decompressLZ(u8** dest, unsigned* dest_sz, const void* src, const size_t s
 
 unsigned _readLZ(int hf, void*& d, unsigned size) {
     // Read file in memory
-    u8* data = (u8*)xr_malloc(size);
+    u8* data = new u8[size / sizeof(u8)];
```
Comment (Im-dex) [inline code/engine/xrCore/LzHuf.cpp]: redundant division

File: code/engine/xrCore/os_clipboard.cpp [hunk 3]
```diff
@@ -69,18 +69,12 @@ void os_clipboard::update_clipboard(LPCSTR string) {
     int memory_length = (int)strlen(memory);
     int string_length = (int)strlen(string);
     int buffer_size = (memory_length + string_length + 1) * sizeof(char);
-#ifndef _EDITOR
-    LPSTR buffer = (LPSTR)_alloca(buffer_size);
-#else  // #ifndef _EDITOR
-    LPSTR buffer = (LPSTR)xr_alloc<char>(buffer_size);
-#endif // #ifndef _EDITOR
+    LPSTR buffer = new char[buffer_size];
```
Comment (Im-dex) [inline code/engine/xrCore/os_clipboard.cpp]: It will be better to  keep _alloca

File: code/engine/xrEngine/ai_script_lua_extension.cpp [hunk 4]
```diff
@@ -236,19 +236,15 @@ bool Script::bfLoadBuffer(CLuaVirtualMachine* tpLuaVM, LPCSTR caBuffer, size_t t
...
         xr_strcpy(script, str_len + tSize, insert);
         std::memcpy(script + str_len, caBuffer, u32(tSize));
         l_iErrorCode = luaL_loadbuffer(tpLuaVM, script, tSize + str_len, caScriptName);
-        xr_free(script);
+        delete[] script;
     } else
         l_iErrorCode = luaL_loadbuffer(tpLuaVM, caBuffer, tSize, caScriptName);
 
     if (l_iErrorCode) {
-#ifdef DEBUG
```
Comment (Im-dex) [inline code/engine/xrEngine/ai_script_lua_extension.cpp]: What for?

File: code/engine/xrEngine/tntQAVI.cpp [hunk 5]
```diff
@@ -192,7 +191,7 @@ BOOL CAviPlayerCustom::Load(char* fname) {
     mmioSeek(hmmioFile, mmckinfoSubchunk.dwDataOffset, SEEK_SET);
 
     // �������� ������ ��� ������  ������ ����� �����
-    m_pMovieData = (BYTE*)xr_malloc(mmckinfoSubchunk.cksize);
+    m_pMovieData = new BYTE[mmckinfoSubchunk.cksize / sizeof(BYTE)];
```
Comment (Im-dex) [inline code/engine/xrEngine/tntQAVI.cpp]: redundant division

File: code/engine/xrEngine/xrImage_Resampler.cpp [hunk 6]
```diff
@@ -28,8 +28,8 @@ Image* new_image(int xsize, int ysize) /* create a blank image */
 {
     Image* image;
 
-    if ((0 != (image = (Image*)xr_malloc(sizeof(Image)))) &&
-        (0 != (image->data = (Pixel*)xr_malloc(ysize * xsize * sizeof(Pixel))))) {
+    if ((0 != (image = new Image)) &&
```
Comment (Im-dex) [inline code/engine/xrEngine/xrImage_Resampler.cpp]: new never returns null

File: code/engine/xrServerEntities/script_storage.cpp [hunk 7]
```diff
@@ -422,34 +416,22 @@ bool CScriptStorage::load_buffer(lua_State* L, LPCSTR caBuffer, size_t tSize, LP
...
-            else {
-#ifdef DEBUG
-                script = (LPSTR)xr_malloc(total_size, "lua script file");
-#else  //#ifdef DEBUG
-                script = (LPSTR)xr_malloc(total_size);
-#endif //#ifdef DEBUG
-                dynamic_allocation = true;
-            }
+                script = new char[total_size / sizeof(char)];
         } __except (GetExceptionCode() == STATUS_STACK_OVERFLOW) {
```
Comment (Im-dex) [inline code/engine/xrServerEntities/script_storage.cpp]: remove stackoverflow catch and reset, cause there is no alloca anymore

File: code/engine/xrEngine/tntQAVI.cpp [hunk 8]
```diff
@@ -114,7 +113,7 @@ BOOL CAviPlayerCustom::Load(char* fname) {
 
     R_ASSERT(m_dwWidth && m_dwHeight);
 
-    m_pDecompressedBuf = (BYTE*)xr_malloc(m_dwWidth * m_dwHeight * 4 + 4);
+    m_pDecompressedBuf = new BYTE[m_dwWidth * m_dwHeight + 1];
```
Comment (Im-dex) [inline code/engine/xrEngine/tntQAVI.cpp]: wrong size

File: code/engine/xrCore/_stl_extensions.h [hunk 9]
```diff
@@ -88,15 +88,8 @@ xalloc<Tp2> __stl_alloc_create(xalloc<Tp1>&, const Tp2*) {
...
-// string(char)
-using xr_string = std::basic_string<char, std::char_traits<char>, xalloc<char>>;
-
-// vector
-template <typename T, typename allocator = xalloc<T>>
-using xr_vector = std::vector<T, allocator>;
-
-template <typename T, typename Alloc>
-void clearAndReserve(xr_vector<T, Alloc>& vec) {
+template <typename T>
```
Comment (Im-dex) [inline code/engine/xrCore/_stl_extensions.h]: Don't remove allocator

File: code/utils/xrDO_Light/StdAfx.h [hunk 10]
```diff
@@ -41,23 +41,23 @@
 #undef THROW
 
 #if XRAY_EXCEPTIONS
-IC xr_string string2xr_string(LPCSTR s) { return s ? s : ""; }
+IC std::string string2std::string(LPCSTR s) { return s ? s : ""; }
```
Comment (Im-dex) [inline code/utils/xrDO_Light/StdAfx.h]: invalid

**Review by Im-dex — CHANGES_REQUESTED** [review]

File: code/utils/xrDO_Light/StdAfx.h [hunk 11]
```diff
@@ -41,23 +41,22 @@
 #undef THROW
 
 #if XRAY_EXCEPTIONS
-IC xr_string string2xr_string(LPCSTR s) { return s ? s : ""; }
 #define THROW(xpr)                           \
     if (!(xpr)) {                            \
         throw __FILE__LINE__ "\"" #xpr "\""; \
     }
 #define THROW2(xpr, msg0)                                                                          \
```
Comment (Im-dex) [inline code/utils/xrDO_Light/StdAfx.h:48]: c_str usage is redundant

File: code/utils/xrDO_Light/StdAfx.h [hunk 12]
```diff
@@ -41,23 +41,22 @@
...
     if (!(xpr)) {                                                                                  \
-        throw xr_string(__FILE__LINE__).append(" \"").append(#xpr).append(string2xr_string(msg0)); \
+        throw std::string(__FILE__LINE__).append(" \"").append(#xpr).append(std::string(msg0).c_str()); \
     }
 #define THROW3(xpr, msg0, msg1)              \
```
Comment (Im-dex) [inline code/utils/xrDO_Light/StdAfx.h]: c_str usage is redundant

File: code/utils/xrAI/StdAfx.h [hunk 13]
```diff
@@ -41,27 +41,26 @@
...
 
 #if XRAY_EXCEPTIONS
-IC xr_string string2xr_string(LPCSTR s) { return s ? s : ""; }
 #define THROW(xpr)                           \
     if (!(xpr)) {                            \
         throw __FILE__LINE__ "\"" #xpr "\""; \
     }
 #define THROW2(xpr, msg0)                                      \
     if (!(xpr)) {                                              \
-        throw * shared_str(xr_string(__FILE__LINE__)           \
```
Comment (Im-dex) [inline code/utils/xrAI/StdAfx.h:51]: c_str usage is redundant

File: code/utils/xrAI/StdAfx.h [hunk 14]
```diff
@@ -41,27 +41,26 @@
...
+        throw * shared_str(std::string(__FILE__LINE__)           \
                                .append(" \"")                  \
                                .append(#xpr)                   \
-                               .append(string2xr_string(msg0)) \
+                               .append(std::string(msg0).c_str()) \
                                .c_str());                      \
     }
 #define THROW3(xpr, msg0, msg1)                                \
```
Comment (Im-dex) [inline code/utils/xrAI/StdAfx.h]: c_str usage is redundant

**Review by Im-dex — CHANGES_REQUESTED** [review]

File: code/utils/xrDO_Light/StdAfx.h (hunk 11 above)
Comment (jazzvaz) [reply code/utils/xrDO_Light/StdAfx.h:48]: fixed

**Review by Im-dex — APPROVED** [review]
Thanks

**Pull request merged**